astropy/utils/shapes.py

```diff
@@ -204,28 +223,28 @@ def _apply(method, *args, **kwargs):
         """
 
     @property
-    def ndim(self):
+    def ndim(self) -> int:
         """The number of dimensions of the instance and underlying arrays."""
         return len(self.shape)
 
     @property
-    def size(self):
+    def size(self) -> int:
         """The size of the object, as calculated from its shape."""
         size = 1
         for sh in self.shape:
             size *= sh
         return size
```

Comment:
Unrelated to this PR. This code can be optimized!

Reply:
There was discussion about it at the time, and timing experiments suggested this was close to optimal already (certainly, `np.prod(self.shape)` is far slower).

Reply:
``math.prod`` would be faster and simpler than the current code:

```python
In [1]: from math import prod

In [2]: import numpy as np

In [3]: def size(shape):
   ...:     size = 1
   ...:     for sh in shape:
   ...:         size *= sh
   ...:     return size
   ...: 

In [4]: shape = (10, 20, 22)

In [5]: %timeit np.prod(shape)
2.17 µs ± 12.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [6]: %timeit prod(shape)
57.7 ns ± 1.07 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [7]: %timeit size(shape)
81.6 ns ± 0.289 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Reply:
Not that 30ns is going to matter much! But at least the code would be simpler.

Reply:
Yes, good to know about `math.prod`. It was introduced in 3.8, so no surprise it wasn't used before. And we can safely use it now.

Reply:
Definitely, and I didn't mean to suggest otherwise. I just wanted to point to a similar discussion I had some time ago :)